contracts/src/Gateway.sol

```diff
@@ -617,4 +621,31 @@ contract Gateway is IGateway, IInitializable {
         assets.assetHubCreateAssetFee = config.assetHubCreateAssetFee;
         assets.assetHubReserveTransferFee = config.assetHubReserveTransferFee;
     }
+
+    /// @inheritdoc IGateway
+    function transact(ParaID destinationChain, TransactMessage calldata message) external payable {
+        Ticket memory ticket;
+        Costs memory costs;
+        address sender;
+        bytes1 originKind;
+        // Mapping originKind to https://github.com/Snowfork/polkadot-sdk/blob/348a1a010481002e41594ed75e5d78b7c2dbed92/polkadot/xcm/src/v2/mod.rs#L86
+        // only support originKind as SovereignAccount or Xcm for now
+        // for Xcm the sender will be the agent of the channel which to construct `DescendOrigin` on BH
+        if (message.originKind == OriginKind.SovereignAccount) {
+            sender = msg.sender;
+            originKind = 0x01;
+        } else if (message.originKind == OriginKind.Xcm) {
+            Channel storage channel = _ensureChannel(destinationChain.into());
+            sender = channel.agent;
+            originKind = 0x03;
+            costs.foreign = message.fee;
+        } else {
+            revert InvalidTransact();
+        }
```

Comment:
I think there is a misunderstanding here. In both cases (OriginKind.SovereignAccount or OriginKind.Xcm), the sender should be `msg.sender`. We don't want agents to be marked as senders.

Consider the following flow:

1. Message reaches bridgeHub, and we translate it to the following:
```
UniversalOrigin(GlobalConsensus(Ethereum)),
DescendOrigin(AccountKey20 { msg.sender })
Transact { OriginKind = "SovereignAccount" OR "Xcm" }
```

2. On the destination parachain, the following converter converts the XCM origin (a MultiLocation) based on the following rules:
https://github.com/Snowfork/polkadot-sdk/blob/c23c50f2cb4697885e1dd891da6fdd33036631a8/cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/xcm_config.rs#L114

3. If `OriginKind` is `Xcm`, then the `XcmPassthrough<RuntimeOrigin>` converter is used, which just passes the multilocation through unmodified.

4. If `OriginKind` is `SovereignAccount`, then the `SovereignSignedViaLocation<LocationToAccountId, RuntimeOrigin>` converter is used, I think. The dest must make sure that `LocationToAccountId` can handle AccountKey20 junctions.

So we can see from this flow that we don't need to involve agent addresses or anything like that.

Reply:
That's true. With `dest must make sure that LocationToAccountId can handle AccountKey20 junctions` it's possible to use the original `msg.sender` in both cases. 

I prefer to keep the agent mode for reasons:

- For some extrinsics the dest chain may not want any `SovereignAccount` to invoke but only allowed from a specific Origin, something like the [CreateOrigin](https://github.com/Snowfork/polkadot-sdk/blob/348a1a010481002e41594ed75e5d78b7c2dbed92/cumulus/parachains/runtimes/assets/asset-hub-rococo/src/lib.rs#L415C2-L415C38)  filter when register foreign token on AH.
- For `SovereignAccount` we don't charge them on Ethereum side for the foreign/xcm fee part by default. Instead we'll require end user to manually prefund his derived account(a keyless Address32 mapping to the ethereum address) on dest chain before they make the call.  It's always a bit challege. While for Agent mode by setting `costs.foreign = message.fee` we charge fees on Etherum on their behalf and use that to refund the Agent sovereign account on dest chain(without requiring end user to do that by themselves).

Reply:
Btw:
```
UniversalOrigin(GlobalConsensus(Ethereum)),
DescendOrigin(AccountKey20 { msg.sender })
Transact { OriginKind = "SovereignAccount" OR "Xcm" }
```
won't not work without fees, my estimated Xcm(not tested yet) will be something like

```
UniversalOrigin(GlobalConsensus(Ethereum)),
DescendOrigin(AccountKey20 { msg.sender }),
WithdrawAsset((Location::parent(),fee)),
BuyExecution { fees: (Location::parent(),fee), weight_limit: Unlimited }
Transact { OriginKind = "SovereignAccount" OR "Xcm" }
```

Reply:
> For Xcm mode there are extrinsics with high security the dest chain won't like to expose to any EOA on Ethereum. Instead only allow to invoke from a specific xcm origin(e.g. only from it's own agent through our bridge), something like the CreateOrigin filter on AH but more strict.

I don't quite agree. With Xcm mode, the extrinsic on the destination parachain can have an origin check that verifies the origin location `../AccountKey20 { }`.

This can prevent unauthorized EOA's from calling the extrinsic.

Reply:
Ok, I'll use `msg.sender` in both cases.